File: LeetCode/Trees/ValidateBST.js
/**
 * Definition for a binary tree node.
 * 
 * The number of nodes in the tree is in the range [1, 10^4]
 * -2^31 <= Node.val <= 2^31 - 1
 * 
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {boolean}
 */

var validate = function(root, low, high) {

    if (!root) return true

    if ((low && root.val <= low) || (high && root.val >= high)) return false

    return validate(root.right, root.val, high) && validate(root.left, low, root.val)
}

var isValidBST = function(root) {
   return validate(root, null, null)
}

console.log(isValidBST([2,1,3]))
console.log(isValidBST([5,1,4,null,null,3,6]))

File: LeetCode/Trees/ValidateBST.cpp
#include <iostream>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  bool isValidBST(TreeNode *root)
  {
    return true;
  }

  void buildTreeFromArray(TreeNode *root, int *arr)
  {
    if (root == nullptr)
    {
      return;
    }
    root->val = *arr;
    buildTreeFromArray(root->left, arr + 1);
    buildTreeFromArray(root->right, arr + 1);
  }
};

int main()
{
  TreeNode *root = new TreeNode();
  Solution *s = new Solution();
  int arr[] = {2, 1, 3};
  s->buildTreeFromArray(root, arr);
  return s->isValidBST(root) ? 0 : 1;
}